pkg/ipam/poolallocator/allocator.go

```diff
@@ -0,0 +1,254 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package poolallocator
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	crdclientset "antrea.io/antrea/pkg/client/clientset/versioned"
+	"antrea.io/antrea/pkg/ipam/ipallocator"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+)
+
+// IPPoolAllocator is responsible for allocating IPs from IP set defined in IPPool CRD.
+// The will update CRD usage accordingly.
+type IPPoolAllocator struct {
+	// Name of IP Pool custom resource
+	ipPoolName string
+
+	// crd client to access the pool
+	crdClient crdclientset.Interface
+}
+
+// NewIPPoolAllocator creates an IPPoolAllocator based on the provided IP pool.
+func NewIPPoolAllocator(poolName string, client crdclientset.Interface) (*IPPoolAllocator, error) {
+
+	allocator := &IPPoolAllocator{
+		ipPoolName: poolName,
+		crdClient:  client,
+	}
+	return allocator, nil
+}
+
+// initAllocatorList reads IP Pool status and initializes a list of allocators based on
+// IP Pool spec and state of allocation recorded in the status
+func (a *IPPoolAllocator) initIPAllocators(ipPool *v1alpha2.IPPool) (ipallocator.MultiIPAllocator, error) {
+
+	var allocators ipallocator.MultiIPAllocator
+
+	// Initialize a list of IP allocators based on pool spec
+	for _, ipRange := range ipPool.Spec.IPRanges {
+		if len(ipRange.CIDR) > 0 {
+			allocator, err := ipallocator.NewCIDRAllocator(ipRange.CIDR)
+			if err != nil {
+				return allocators, err
+			}
+			allocators = append(allocators, allocator)
+		} else {
+			allocator, err := ipallocator.NewIPRangeAllocator(ipRange.Start, ipRange.End)
+			if err != nil {
+				return allocators, err
+			}
+			allocators = append(allocators, allocator)
+		}
+	}
+
+	// Mark allocated IPs from pool status as unavailable
+	for _, ip := range ipPool.Status.Usage {
+		err := allocators.AllocateIP(net.ParseIP(ip.IPAddress))
+		if err != nil {
+			// TODO - fix state if possible
+			return allocators, fmt.Errorf("Inconsistent state for IP Pool %s with IP %s", ipPool.Name, ip.IPAddress)
+		}
+	}
+
+	return allocators, nil
+}
+
+func (a *IPPoolAllocator) readPoolAndInitIPAllocators() (*v1alpha2.IPPool, ipallocator.MultiIPAllocator, error) {
+	ipPool, err := a.crdClient.CrdV1alpha2().IPPools().Get(context.TODO(), a.ipPoolName, metav1.GetOptions{})
```

Comment:
Do you think we can use informer for this?

Reply:
Hi Ran, what would be the advantages of using informer here?

Reply:
informer is a memory cache of K8s resources. If we get too many requests, informer is better than simple client.
Note: do not add PodInformer to antrea-agent since this may cost too much memory.

Reply:
I think it's worth using it. We do not expect many pool objects - and they should be fairly static.